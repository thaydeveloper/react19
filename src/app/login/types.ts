import { FormEvent } from "react";

export interface LoginPresentationProps {
  email: string;
  password: string;
  onEmailChange: (value: string) => void;
  onPasswordChange: (value: string) => void;
  onSubmit: (e: FormEvent<HTMLFormElement>) => void;
}
