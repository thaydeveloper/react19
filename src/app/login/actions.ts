"use server";

import { AuthService } from "@/api/auth.service";
import { LoginState } from "@/interfaces/login.interfaces";

export async function loginAction(
  prevState: LoginState,
  formData: FormData
): Promise<LoginState> {
  try {
    const email = formData.get("email");
    const password = formData.get("password");

    if (!email || !password) {
      return {
        success: false,
        message: "Email e senha são obrigatórios",
        error: null,
      };
    }

    const result = await AuthService.login({
      email: email as string,
      password: password as string,
    });
    console.log("Resposta recebida: no action", result);

    return {
      success: true,
      message: "Login realizado com sucesso",
      data: result,
    };
  } catch (error: any) {
    const errorMessage =
      error.response?.data?.message ||
      error.message ||
      "Erro ao realizar login";

    console.error("LoginAction - Erro completo:", {
      message: errorMessage,
      status: error.response?.status,
      data: error.response?.data,
    });

    return {
      success: false,
      message: errorMessage,
      error: error as Error,
    };
  }
}
