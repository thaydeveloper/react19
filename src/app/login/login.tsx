'use client';

import React from 'react';

interface LoginPresentationProps {
  formAction: (formData: FormData) => void;
  isSubmitting: boolean;
  error: string | null;
}

export function LoginPresentation({
  formAction,
  isSubmitting,
  error,
}: LoginPresentationProps) {
  return (
    <div className="flex min-h-screen items-center justify-center bg-gray-100 p-4">
      <div className="w-full max-w-md rounded-lg bg-white p-8 shadow-lg">
        <h2 className="mb-6 text-center text-3xl font-bold text-gray-900">
          Login
        </h2>

        {error && (
          <div
            className="mb-4 rounded-md bg-red-50 p-4 text-red-700"
            role="alert"
            data-testid="error-message"
          >
            {error}
          </div>
        )}

        <form
          className="space-y-4"
          action={formAction}
          data-testid="login-form"
        >
          <div>
            <label
              htmlFor="email"
              className="block text-sm font-medium text-gray-700"
            >
              Email
            </label>
            <input
              type="email"
              id="email"
              name="email"
              required
              placeholder="Digite seu email"
              className="mt-2 block w-full h-[50px] p-[10px] rounded-md border-gray-300 shadow-sm focus:border-blue-500 focus:ring-blue-500"
            />
          </div>
          <div>
            <label
              htmlFor="password"
              className="block text-sm font-medium text-gray-700"
            >
              Senha
            </label>
            <input
              type="password"
              id="password"
              name="password"
              required
              placeholder="Digite sua senha"
              className="mt-1 block w-full h-[50px] p-[10px] rounded-md border-gray-300 shadow-sm focus:border-blue-500 focus:ring-blue-500"
            />
          </div>
          <button
            type="submit"
            disabled={isSubmitting}
            className="mt-4 w-full rounded-md bg-blue-600 py-2 px-4 text-white hover:bg-blue-700 focus:outline-none focus:ring-2 focus:ring-blue-500 disabled:bg-blue-300"
          >
            {isSubmitting ? 'Entrando...' : 'Entrar'}
          </button>
        </form>
      </div>
    </div>
  );
}
