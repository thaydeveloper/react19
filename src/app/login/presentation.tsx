import React from "react";
import { LoginPresentationProps } from "./types";

export const LoginPresentation: React.FC<LoginPresentationProps> = ({
  email,
  password,
  onEmailChange,
  onPasswordChange,
  onSubmit,
}) => {
  return (
    <div className="flex min-h-screen items-center justify-center bg-gray-100 p-4">
      <div className="w-full max-w-md rounded-lg bg-white p-8 shadow-lg">
        <h2 className="mb-6 text-center text-3xl font-bold text-gray-900">
          Login
        </h2>
        <form onSubmit={onSubmit} className="space-y-4">
          <div>
            <label
              htmlFor="email"
              className="block text-sm font-medium text-gray-700"
            >
              Email
            </label>
            <input
              type="email"
              id="email"
              value={email}
              onChange={(e) => onEmailChange(e.target.value)}
              placeholder="Digite seu email"
              className="mt-1 block w-full rounded-md border-gray-300 shadow-sm focus:border-blue-500 focus:ring-blue-500"
              required
            />
          </div>
          <div>
            <label
              htmlFor="password"
              className="block text-sm font-medium text-gray-700"
            >
              Senha
            </label>
            <input
              type="password"
              id="password"
              value={password}
              onChange={(e) => onPasswordChange(e.target.value)}
              placeholder="Digite sua senha"
              className="mt-1 block w-full rounded-md border-gray-300 shadow-sm focus:border-blue-500 focus:ring-blue-500"
              required
            />
          </div>
          <button
            type="submit"
            className="mt-4 w-full rounded-md bg-blue-600 py-2 px-4 text-white hover:bg-blue-700 focus:outline-none focus:ring-2 focus:ring-blue-500"
          >
            Entrar
          </button>
        </form>
      </div>
    </div>
  );
};
