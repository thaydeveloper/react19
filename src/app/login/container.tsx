'use client';

import { useOptimistic, useActionState } from 'react';
import { LoginPresentation } from './login';
import { loginAction } from './actions';
import { initialState } from '../../constants/login.constants';
import { OptimisticState } from '../../interfaces/login.interfaces';
import { useRouter } from 'next/navigation';

export function LoginContainer() {
  const router = useRouter();
  const [state, , isPending] = useActionState(loginAction, initialState);
  const [optimisticState, addOptimistic] = useOptimistic(
    { isSubmitting: false },
    (state: OptimisticState, newState: Partial<OptimisticState>) => ({
      ...state,
      ...newState,
    }),
  );

  async function handleSubmit(formData: FormData) {
    addOptimistic({ isSubmitting: true });

    try {
      const result = await loginAction(initialState, formData);

      if (result?.success && result.data) {
        localStorage.setItem('token', result.data.token);
        document.cookie = `auth_token=${result.data.token}; path=/`;
        router.push('/dashboard');
      }
    } catch (error) {
      console.error('Erro no login:', error);
    } finally {
      addOptimistic({ isSubmitting: false });
    }
  }

  return (
    <LoginPresentation
      formAction={handleSubmit}
      isSubmitting={optimisticState.isSubmitting || isPending}
      error={!state.success ? state.message : null}
    />
  );
}
