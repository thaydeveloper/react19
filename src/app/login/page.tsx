import { LoginContainer } from "./container";
import React from "react";

export default function LoginPage() {
  return <LoginContainer />;
}
