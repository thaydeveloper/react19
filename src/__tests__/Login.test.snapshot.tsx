import React from 'react';
import { render } from '@testing-library/react';
import { LoginPresentation } from '../app/login/login';

describe('LoginPresentation Snapshots', () => {
  const mockFormAction = jest.fn();

  test('deve corresponder ao snapshot no estado inicial', () => {
    const { container } = render(
      <LoginPresentation formAction={mockFormAction} isSubmitting={false} />,
    );
    expect(container).toMatchSnapshot();
  });

  test('deve corresponder ao snapshot no estado de carregamento', () => {
    const { container } = render(
      <LoginPresentation formAction={mockFormAction} isSubmitting={true} />,
    );
    expect(container).toMatchSnapshot();
  });

  test('deve corresponder ao snapshot quando há erro', () => {
    const { container } = render(
      <LoginPresentation
        formAction={mockFormAction}
        isSubmitting={false}
        error="Email ou senha inválidos"
      />,
    );
    // Após as correções, execute 'yarn test -u' para atualizar os snapshots
    expect(container).toMatchSnapshot();
  });
});

// Após fazer as alterações, execute o comando para atualizar os snapshots:
// yarn test -u
