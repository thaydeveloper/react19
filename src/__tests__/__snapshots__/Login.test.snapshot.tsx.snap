// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginPresentation Snapshots deve corresponder ao snapshot no estado de carregamento 1`] = `
<div>
  <div
    class="flex min-h-screen items-center justify-center bg-gray-100 p-4"
  >
    <div
      class="w-full max-w-md rounded-lg bg-white p-8 shadow-lg"
    >
      <h2
        class="mb-6 text-center text-3xl font-bold text-gray-900"
      >
        Login
      </h2>
      <form
        action="javascript:throw new Error('A React form was unexpectedly submitted. If you called form.submit() manually, consider using form.requestSubmit() instead. If you\\'re trying to use event.stopPropagation() in a submit event handler, consider also calling event.preventDefault().')"
        class="space-y-4"
        data-testid="login-form"
      >
        <div>
          <label
            class="block text-sm font-medium text-gray-700"
            for="email"
          >
            Email
          </label>
          <input
            class="mt-2 block w-full h-[50px] p-[10px] rounded-md border-gray-300 shadow-sm focus:border-blue-500 focus:ring-blue-500"
            id="email"
            name="email"
            placeholder="Digite seu email"
            required=""
            type="email"
          />
        </div>
        <div>
          <label
            class="block text-sm font-medium text-gray-700"
            for="password"
          >
            Senha
          </label>
          <input
            class="mt-1 block w-full h-[50px] p-[10px] rounded-md border-gray-300 shadow-sm focus:border-blue-500 focus:ring-blue-500"
            id="password"
            name="password"
            placeholder="Digite sua senha"
            required=""
            type="password"
          />
        </div>
        <button
          class="mt-4 w-full rounded-md bg-blue-600 py-2 px-4 text-white hover:bg-blue-700 focus:outline-none focus:ring-2 focus:ring-blue-500 disabled:bg-blue-300"
          disabled=""
          type="submit"
        >
          Entrando...
        </button>
      </form>
    </div>
  </div>
</div>
`;

exports[`LoginPresentation Snapshots deve corresponder ao snapshot no estado inicial 1`] = `
<div>
  <div
    class="flex min-h-screen items-center justify-center bg-gray-100 p-4"
  >
    <div
      class="w-full max-w-md rounded-lg bg-white p-8 shadow-lg"
    >
      <h2
        class="mb-6 text-center text-3xl font-bold text-gray-900"
      >
        Login
      </h2>
      <form
        action="javascript:throw new Error('A React form was unexpectedly submitted. If you called form.submit() manually, consider using form.requestSubmit() instead. If you\\'re trying to use event.stopPropagation() in a submit event handler, consider also calling event.preventDefault().')"
        class="space-y-4"
        data-testid="login-form"
      >
        <div>
          <label
            class="block text-sm font-medium text-gray-700"
            for="email"
          >
            Email
          </label>
          <input
            class="mt-2 block w-full h-[50px] p-[10px] rounded-md border-gray-300 shadow-sm focus:border-blue-500 focus:ring-blue-500"
            id="email"
            name="email"
            placeholder="Digite seu email"
            required=""
            type="email"
          />
        </div>
        <div>
          <label
            class="block text-sm font-medium text-gray-700"
            for="password"
          >
            Senha
          </label>
          <input
            class="mt-1 block w-full h-[50px] p-[10px] rounded-md border-gray-300 shadow-sm focus:border-blue-500 focus:ring-blue-500"
            id="password"
            name="password"
            placeholder="Digite sua senha"
            required=""
            type="password"
          />
        </div>
        <button
          class="mt-4 w-full rounded-md bg-blue-600 py-2 px-4 text-white hover:bg-blue-700 focus:outline-none focus:ring-2 focus:ring-blue-500 disabled:bg-blue-300"
          type="submit"
        >
          Entrar
        </button>
      </form>
    </div>
  </div>
</div>
`;

exports[`LoginPresentation Snapshots deve corresponder ao snapshot quando há erro 1`] = `
<div>
  <div
    class="flex min-h-screen items-center justify-center bg-gray-100 p-4"
  >
    <div
      class="w-full max-w-md rounded-lg bg-white p-8 shadow-lg"
    >
      <h2
        class="mb-6 text-center text-3xl font-bold text-gray-900"
      >
        Login
      </h2>
      <div
        class="mb-4 rounded-md bg-red-50 p-4 text-red-700"
        data-testid="error-message"
        role="alert"
      >
        Email ou senha inválidos
      </div>
      <form
        action="javascript:throw new Error('A React form was unexpectedly submitted. If you called form.submit() manually, consider using form.requestSubmit() instead. If you\\'re trying to use event.stopPropagation() in a submit event handler, consider also calling event.preventDefault().')"
        class="space-y-4"
        data-testid="login-form"
      >
        <div>
          <label
            class="block text-sm font-medium text-gray-700"
            for="email"
          >
            Email
          </label>
          <input
            class="mt-2 block w-full h-[50px] p-[10px] rounded-md border-gray-300 shadow-sm focus:border-blue-500 focus:ring-blue-500"
            id="email"
            name="email"
            placeholder="Digite seu email"
            required=""
            type="email"
          />
        </div>
        <div>
          <label
            class="block text-sm font-medium text-gray-700"
            for="password"
          >
            Senha
          </label>
          <input
            class="mt-1 block w-full h-[50px] p-[10px] rounded-md border-gray-300 shadow-sm focus:border-blue-500 focus:ring-blue-500"
            id="password"
            name="password"
            placeholder="Digite sua senha"
            required=""
            type="password"
          />
        </div>
        <button
          class="mt-4 w-full rounded-md bg-blue-600 py-2 px-4 text-white hover:bg-blue-700 focus:outline-none focus:ring-2 focus:ring-blue-500 disabled:bg-blue-300"
          type="submit"
        >
          Entrar
        </button>
      </form>
    </div>
  </div>
</div>
`;
